From 5ae476b092245722042d376f94a9da150cbc1790 Mon Sep 17 00:00:00 2001
From: Xi Ruoyao <xry111@xry111.site>
Date: Fri, 15 Aug 2025 23:23:20 +0800
Subject: [PATCH 38/63] loonggpu: Don't directly use ttm_bo_get

Adapt for upstream API change which made this function internal.

Link: https://git.kernel.org/torvalds/c/9ec1ac835e48
Signed-off-by: Xi Ruoyao <xry111@xry111.site>
---
 include/loonggpu.h         | 2 ++
 loonggpu/loonggpu_gem.c    | 7 +------
 loonggpu/loonggpu_object.c | 8 +++-----
 3 files changed, 6 insertions(+), 11 deletions(-)

diff --git a/include/loonggpu.h b/include/loonggpu.h
index d20472b..6cec110 100644
--- a/include/loonggpu.h
+++ b/include/loonggpu.h
@@ -308,6 +308,8 @@ struct loonggpu_cs_post_dep {
 #define gem_to_loonggpu_bo(gobj) container_of((gobj), struct loonggpu_bo, gem_base)
 #endif
 
+extern const struct drm_gem_object_funcs loonggpu_gem_object_funcs;
+
 void loonggpu_gem_object_free(struct drm_gem_object *obj);
 int loonggpu_gem_object_open(struct drm_gem_object *obj,
 				struct drm_file *file_priv);
diff --git a/loonggpu/loonggpu_gem.c b/loonggpu/loonggpu_gem.c
index 05b240c..b441934 100644
--- a/loonggpu/loonggpu_gem.c
+++ b/loonggpu/loonggpu_gem.c
@@ -81,11 +81,10 @@ void loonggpu_gem_object_free(struct drm_gem_object *gobj)
 
 	if (robj) {
 		loonggpu_mn_unregister(robj);
-		loonggpu_bo_unref(&robj);
+		ttm_bo_put(&robj->tbo);
 	}
 }
 
-#if !defined(LG_DRM_DRIVER_HAS_GEM_FREE)
 const struct drm_gem_object_funcs loonggpu_gem_object_funcs = {
 	.free = loonggpu_gem_object_free,
 	.open = loonggpu_gem_object_open,
@@ -96,7 +95,6 @@ const struct drm_gem_object_funcs loonggpu_gem_object_funcs = {
 	.mmap = loonggpu_gem_object_mmap,
 	.vm_ops = &loonggpu_gem_vm_ops,
 };
-#endif
 
 int loonggpu_gem_object_create(struct loonggpu_device *adev, unsigned long size,
 			     int alignment, u32 initial_domain,
@@ -142,9 +140,6 @@ retry:
 		return r;
 	}
 	*obj = &lg_gbo_to_gem_obj(bo);
-#if !defined(LG_DRM_DRIVER_HAS_GEM_FREE)
-	(*obj)->funcs = &loonggpu_gem_object_funcs;
-#endif
 	return 0;
 }
 
diff --git a/loonggpu/loonggpu_object.c b/loonggpu/loonggpu_object.c
index 73eaf7d..9fb5e78 100644
--- a/loonggpu/loonggpu_object.c
+++ b/loonggpu/loonggpu_object.c
@@ -454,6 +454,7 @@ static int loonggpu_bo_do_create(struct loonggpu_device *adev,
 	if (bo == NULL)
 		return -ENOMEM;
 	drm_gem_private_object_init(adev->ddev, &lg_gbo_to_gem_obj(bo), size);
+	bo->tbo.base.funcs = &loonggpu_gem_object_funcs;
 	INIT_LIST_HEAD(&bo->shadow_list);
 	INIT_LIST_HEAD(&bo->va);
 	bo->preferred_domains = bp->preferred_domain ? bp->preferred_domain :
@@ -811,7 +812,7 @@ struct loonggpu_bo *loonggpu_bo_ref(struct loonggpu_bo *bo)
 	if (bo == NULL)
 		return NULL;
 
-	ttm_bo_get(&bo->tbo);
+	drm_gem_object_get(&bo->tbo.base);
 	return bo;
 }
 
@@ -823,13 +824,10 @@ struct loonggpu_bo *loonggpu_bo_ref(struct loonggpu_bo *bo)
  */
 void loonggpu_bo_unref(struct loonggpu_bo **bo)
 {
-	struct ttm_buffer_object *tbo;
-
 	if ((*bo) == NULL)
 		return;
 
-	tbo = &((*bo)->tbo);
-	ttm_bo_put(tbo);
+	drm_gem_object_put(&(*bo)->tbo.base);
 	*bo = NULL;
 }
 
-- 
2.52.0

